m2m-oauth-server: Managing M2M OAuth Server Tokens

jkralik opened the pull request:
### Overview
This pull request introduces comprehensive functionalities for managing tokens in the M2M OAuth server, enhancing the server's ability to handle token creation, retrieval, blacklisting, and deletion.

### Key Features

**Token Creation:**
- Support for generating access tokens with specific claims (subject, audience, scopes, expiration time).
- Crucial for secure authentication in machine-to-machine (M2M) interactions.

**Token Retrieval:**
- Tokens can be retrieved based on various criteria: owner, token ID, and blacklisted status.
- Facilitates efficient management and monitoring of issued tokens.

**Token Blacklisting:**
- Ability to blacklist tokens, preventing further use.
- Enhances security by identifying and managing compromised or obsolete tokens.

**Token Deletion:**
- Automatic deletion of expired tokens from the store.
- Maintains a clean and up-to-date token database, ensuring optimal performance and security.

**Configuration Enhancements:**
- Updates to configuration structures for more flexible and secure management of OAuth clients and their configurations.

**Protocol Buffers and gRPC:**
- Introduction of Protocol Buffers definitions and gRPC services for token management operations.
- Facilitates efficient communication and integration with other services.
- Definitions include messages for creating, retrieving, and blacklisting tokens, along with HTTP and OpenAPI annotations for RESTful API generation.

**Documentation:**
- Detailed documentation in Markdown and HTML formats.
- Describes the protocol, message types, and service methods.
- Ensures developers can easily understand and integrate the new functionalities.

**Web interface:**
- Users are able to manage their tokens.

### Benefits
These enhancements make the M2M OAuth server more robust and secure, providing essential features for managing API access tokens in a machine-to-machine context.

jkralik committed:
m2m-oauth-server: define RPC calls

jkralik committed:
implement mongodb

jkralik committed:
update swagger

jkralik committed:
add time_to_live in ns

jkralik committed:
m2m-oauth-server supports grpc and http APIs

jkralik committed:
store set index

jkralik committed:
fix linter

jkralik committed:
set helm charts

jkralik committed:
fix chart

jkralik committed:
fix validator

jkralik committed:
cleanup

jkralik committed:
fix ports in helm

jkralik committed:
fix deployment

jkralik committed:
use seconds instead of nanoseconds in token from grpc

jkralik committed:
fix sonarcloud issue

Danielius1922 commented on m2m-oauth-server/oauthSigner/oauthSigner.go:
```diff
@@ -0,0 +1,99 @@
+type OAuthSigner struct {
+	privateKeyJWTValidators map[string]*validator.Validator
+	closer                  fn.FuncList
+	Config                  Config
```
Do we want to allow other packages to modify the Config? If not, the I would go for unexported config and a getter

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/pb/tags.go:
```diff
@@ -0,0 +1,11 @@
+const (
+	ExpirationKey           = "expiration"
+	OwnerKey                = "owner"
+	BlackListedFlagKey      = "blacklisted.flag"
```
BlackListedKey + ".flag" ?

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/service/grpc/service.go:
```diff
@@ -0,0 +1,35 @@
+	}
+	pb.RegisterM2MOAuthServiceServer(server.Server, m2mOAuthServiceServer)
+
+	// SnippetService needs to stop gracefully to ensure that all commands are processed.
```
M2MOAuthService

jkralik replied:
fixed

jkralik committed:
add subject

jkralik committed:
add subject

jkralik commented:
@coderabbitai generate a description for this pullrequest about the managing m2m-oauth-server tokens feature targeting to not plgd developers.

jkralik committed:
fix crt

Danielius1922 commented on charts/plgd-hub/templates/m2m-oauth-server/config.yaml:
```diff
@@ -17,16 +17,49 @@ data:
+          # 0s - means infinity
+          maxConnectionAgeGrace: {{ .apis.grpc.keepAlive.maxConnectionAgeGrace }}
+          time: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
+          timeout: {{ .apis.grpc.keepAlive.maxConnectionIdle }}
```
`maxConnectionAge`, `time` and `timeout` are all set to `{{ .apis.grpc.keepAlive.maxConnectionIdle }}` is that correct?

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/store/store.go:
```diff
@@ -0,0 +1,76 @@
+	ErrPartialDelete   = errors.New("some errors occurred while deleting")
+)
+
+func IsDuplicateKeyError(err error) bool {
```
rm? doesn't seem to be used

jkralik replied:
removed

Danielius1922 commented on m2m-oauth-server/swagger.yaml:
```diff
@@ -18,6 +31,12 @@ paths:
+                  description: "The name of the token which will be used in the name claim."
+                expiration:
+                  type: integer
+                  description: "The requested expiration time in seconds from the client. If not provided, the token will be max allowed by client."
```
`the token will be max allowed by client` is missing something, maybe - the token will have the max expiration time allowed by the client?

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/service/grpc/token.go:
```diff
@@ -0,0 +1,288 @@
+		return time.Time{}
+	}
+	if tokenReq.CreateTokenRequest.GetExpiration() == 0 {
+		if clientCfg.AccessTokenLifetime == 0 {
```
this will never execute since it is checked already on line 111 and there it will always return from the function

jkralik replied:
removed

Danielius1922 commented on m2m-oauth-server/service/http/requestHandler.go:
```diff
@@ -0,0 +1,46 @@
+	mux                   *runtime.ServeMux
+}
+
+// NewHTTP returns HTTP handler
```
NewHTTP -> NewRequestHandler

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/store/mongodb/tokens_test.go:
```diff
@@ -0,0 +1,293 @@
+	s, cleanUpStore := test.NewMongoStore(t)
+	defer cleanUpStore()
+
+	ctx, cancel := context.WithTimeout(context.Background(), config.TEST_TIMEOUT*100)
```
config.TEST_TIMEOUT

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/service/grpc/token.go:
```diff
@@ -0,0 +1,288 @@
+
+func validateExpiration(clientCfg *oauthsigner.Client, tokenReq *tokenRequest) error {
+	if tokenReq.CreateTokenRequest.GetExpiration() > 0 {
+		if tokenReq.CreateTokenRequest.GetExpiration() < tokenReq.issuedAt.Unix() {
```
add test case with non-zero expiration so we get better code coverage

jkralik replied:
added

jkralik committed:
fix for CR

Danielius1922 commented on m2m-oauth-server/pb/token.go:
```diff
@@ -0,0 +1,190 @@
+	return nil
+}
+
+func (x *Token) ToMap() (map[string]interface{}, error) {
```
The methods below are for encoding to BSON right? 

There are similar pkg methods `UnmarshalProtoBSON`, `MarshalProtoBSON` and `ConvertStringValueToInt64` that I created when I was implementing snippet-service. I think they do almost the same stuff, maybe we can combine them into a single encode/decode-bson API

jkralik replied:
implemented

jkralik committed:
fix sentence

jkralik committed:
fix for CR 2

Danielius1922 commented on m2m-oauth-server/service/service.go:
```diff
@@ -4,82 +4,152 @@ import (
+			}
+		})
+		if err2 != nil {
+			return nil, fmt.Errorf("cannot create scheduler: %w", err2)
```
s.Close() before returning?

jkralik replied:
fixed

jkralik committed:
fix for CR3

jkralik committed:
fix smells

Danielius1922 committed:
Refactor for m2m token validation

Danielius1922 commented on m2m-oauth-server/store/mongodb/tokens.go:
```diff
@@ -0,0 +1,172 @@
+		setIdOwnerHint = false
+	}
+	if len(req.GetAudienceFilter()) > 0 {
+		filter = append(filter, bson.E{Key: pb.AudienceKey, Value: bson.M{mongodb.In: req.GetAudienceFilter()}})
```
add test with audienceFilter (according to SC this code path is not tested)

jkralik replied:
I will remove audience part for now

Danielius1922 commented on pkg/mongodb/marshal.go:
```diff
@@ -2,68 +2,258 @@ package mongodb
+		case map[string]any:
+			var err error
+			current, err = iterateOverMap(curr, permitMissingPaths, part)
+			if err != nil {
```
```go
if err != nil {
  return nil, err
}
if current == nil {
  return nil, nil
}
```

These lines are the same for both cases, if we move them outside of the switch SC should stop complaining about function complexity

jkralik replied:
fixed

Danielius1922 commented on m2m-oauth-server/service/grpc/token.go:
```diff
@@ -0,0 +1,285 @@
+		}
+		return time.Time{}
+	}
+	if tokenReq.CreateTokenRequest.GetExpiration() == 0 {
```
Can we add a test case that goes through lines 117-125? Otherwise we're fine, even if we don't hit 80%, the rest of the code paths are error paths

jkralik replied:
added

jkralik committed:
fix after CR 4

Danielius1922 committed:
security: check whether token is trusted after validation

Danielius1922 committed:
Add token cache

Danielius1922 committed:
validator: add configuration and helm charts

Danielius1922 committed:
Ensure that for parallel token verification only a single requests is
sent

Danielius1922 committed:
Periodically clean-up token cache

Danielius1922 committed:
Fix static analysis issues and add tests

jkralik committed:
fix delete expired tokens from DB

jkralik committed:
rename BlacklistTokens to DeleteTokens

jkralik committed:
fix call

jkralik committed:
use same name as is in proto for originalTokenClaims

jkralik committed:
fixing m2m-oauth-server-blacklist-client (#1355)

PatrikMatiasko committed:
Implement New UI for API Token Service (#1353)

**Token Generation:**
Added a form to create new API tokens, where users can specify a custom name and select an expiration date.
Included client-side validation to ensure that token names are unique and expiration dates are valid.

**Token Listing:**
Implemented a new interface to display a list of existing tokens, including their names, creation dates, expiration dates, and status. Added sorting and filtering capabilities to help users quickly find specific tokens.

**Token Removal:**
Introduced functionality for users to remove tokens directly from the UI.
Added confirmation dialogs to prevent accidental deletion of tokens.

---------

Co-authored-by: Jozef Kralik <jozef.kralik@plgd.dev>
Co-authored-by: Daniel Adam <daniel.adam1922@protonmail.com>

jkralik committed:
Merge branch 'main' into jkralik/feature/m2m-oauth-server-blacklist

PatrikMatiasko committed:
Sidebar visibility by service ( API TOKENS )

PatrikMatiasko committed:
Conditional routing

PatrikMatiasko committed:
Conditional routing

PatrikMatiasko committed:
Conditional routing

jkralik committed:
helm - enable apiTokens

jkralik committed:
update bundle
